File: docs/algorithms/code/algorithms/insertIntoBST.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function insertIntoBST(root: TreeNode | null, val: number): TreeNode | null {
  if (!root) {
    return new TreeNode(val);
  }
  let head = root;

  while (head) {
    if (head.val > val) {
      if (!head.left) {
        head.left = new TreeNode(val);
        break;
      }
      head = head.left;
    }

    if (head.val < val) {
      if (!head.right) {
        head.right = new TreeNode(val);
        break;
      }
      head = head.right;
    }
  }

  return root;
};

File: docs/algorithms/code/algorithms/insertIntoBST.go
// 思路：从根节点开始遍历，如果当前节点为空，就返回新节点，如果当前节点不为空，就比较当前节点和新节点的大小，如果新节点小，就递归调用当前节点的左子树，如果新节点大，就递归调用当前节点的右子树。
// 复杂度分析：时间复杂度 O(h), 空间复杂度 O(1)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func insertIntoBST(root *TreeNode, val int) *TreeNode {
	if root == nil {
		return &TreeNode{Val: val}
	}

	curr := root
	for curr != nil {
		if val < curr.Val {
			if curr.Left == nil {
				curr.Left = &TreeNode{Val: val}
				break
			} else {
				curr = curr.Left
			}
		} else {
			if curr.Right == nil {
				curr.Right = &TreeNode{Val: val}
				break
			} else {
				curr = curr.Right
			}
		}
	}

	return root
}